client-app/src/desktop/tabs/examples/filemanager/FileManagerModel.js

```diff
@@ -0,0 +1,259 @@
+                formData.append(`file-${idx}`, file, file.name);
+            });
+
+            uploadPromise = XH.fetch({
+                url: 'fileManager/upload',
+                method: 'POST',
+                body: formData,
+                headers: new Headers()
+            });
+        }
+
+        const deletes = this.getFilesToDelete();
+        if (deletes.length) {
+            deletePromise = Promise.all(deletes.map(it => {
+                XH.fetchService
+                    .fetchJson({
+                        url: 'fileManager/delete',
+                        params: {filename: it.name}
+                    })
+                    .then(ret => {
+                        if (!ret.success) throw `Unable to delete ${it.name}`;
+                    })
+                    .catchDefault();
+            }));
+        }
+
+        return Promise
+            .all([
+                uploadPromise,
+                deletePromise
+            ])
+            .then(() => {
+                return this.resetAndLoadAsync();
+            })
+            .then(() => {
+                XH.toast({
+                    message: `Files processed successfully: ${uploads.length} uploaded | ${deletes.length} deleted.`
+                });
+            })
+            .linkTo(this.loadMaskModel)
+            .catchDefault();
+    }
+
+    async resetAndLoadAsync() {
+        this.chooserModel.removeAllFiles();
+        this.loadAsync();
+    }
+
+    async downloadSelectedAsync() {
+        if (!this.enableDownload) return;
+
+        const sel = this.gridModel.selectedRecord,
+            response = await XH.fetch({
+                url: 'fileManager/download',
+                params: {filename: sel.name}
+            }).catchDefault();
+
+        const blob = await response.blob();
+        download(blob, sel.name);
+        XH.toast({
+            icon: Icon.download(),
+            message: 'Download complete.'
+        });
+    }
+
+    // Mark already-uploaded file as pending deletion on save, or immediately remove a
+    // not-yet-uploaded file from the chooser.
+    removeFile(file) {
+        const {status} = file;
+
+        if (status == 'Saved') {
+            // Already uploaded record - change status directly in grid store.
+            this.setFileStatus(file.id, 'Pending Delete');
+        } else if (status == 'Pending Upload') {
+            // Ask chooser to remove - reaction on chooser files[] will update grid store.
+            this.chooserModel.removeFileByName(file.name);
+        }
+    }
+
+    // Restore a file marked as "Pending Delete" - i.e. cancel delete request.
+    restoreFile(file) {
+        if (file.status == 'Pending Delete') {
+            this.setFileStatus(file.id, 'Saved');
+        }
+    }
+
+    // Update the status of an existing record, ensuring store reloaded to reflect.
+    // TODO: store recs -> rawData -> manipulate -> reload (is there a better way?)
```

Comment:
@lbwexler this prompted my question over the past weekend re. preferred patterns for ad hoc updating of fields on records.